content/posts/fpcast.md

```diff
@@ -0,0 +1,478 @@
+
+So writing `my_noargs_function2` invokes undefined behavior in the C
+specification. However, most C compilers produce code for most targets that
+works fine. Since most of the time Python packages are compiled in these cases,
+package authors cast function pointers and move on with their lives. The
+existing Python ecosystem is full of these casts all over the place.
+
+### Function pointer casts don't work in WebAssembly
+
+In x86, calling a function pointer is the same process as calling a static
+function call. Arguments and a return address are pushed onto a stack, and a
+jump is performed. The only difference is that in the case of a static function
+call, the target of the jump is known at compile time, whereas in the case of a
+dynamic call, it is determined by a variable.
+
+In WebAssembly, a function pointer is an index into a table of functions. There
+is a separate WebAssembly instruction called `call_indirect` which takes an
+immediate (compile-time determined) argument to indicate what the signature of
+the function being called should be. The WebAssembly runtime checks that the
+function being called has a signature which matches the asserted signature and
+if it doesn't it throws an error.
+
+So we call `my_noargs_function2` through a `call_indirect` instruction which
+asserts that it takes two arguments. The WebAssembly runtime checks and sees
+that in fact, `my_noargs_function2` only expects one argument, and it crashes
+with an "indirect call signature mismatch" error.
+
+## Ways to fix the call signature mismatch errors
+
+### 1. EMULATE_FUNCTION_POINTER_CASTS
+
+The Emscripten compiler comes with a flag that makes it emulate function pointer
+casting. This is what we have been using to resolve the troubles. However, it
+comes with a high cost: it is slow, it uses up large amounts of stack space, and
+it increases code size by a lot. Also, it makes calling a function pointer from
+Javascript inconvenient, and in particular breaks dynamic linking and many
+Emscripten ports which implement C libraries with Javascript code. In upstream
+Emscripten, EMULATE_FUNCTION_POINTER_CASTS doesn't work with dynamic linking at
+all. It was only due to complicated patches contributed by Joe Marshall that we
+were able to use it. 
+
+I will describe later how EMULATE_FUNCTION_POINTER_CASTS works and why it
+imposes such large costs.
+
+### 2. Just patch the packages
+
+We could just go through and patch the Python C packages to fix their function
+declarations. However, there is no easy way to detect these function pointer
+casts (there's no compiler setting that can generate warnings about them). It is
+also very tedious to locate the problematic code. The upstream packages are
+generally happy to accept these patches but it takes time to review them and
+they may not get backported as far as we need them (Pyodide is stuck on pretty
+old versions of some packages). So it generates a high maintenance burden. This
```

Comment:
I think apart from scipy and dependent packages we are relatively up to date?  Might be worth either clarifying why we are stuck to old versions or removing the part in "(.. )".

Reply:
I will trim a bit here, it's a bit of an aside of an aside.